### PR TITLE
Fix the bug of BidirectionalCell

### PR DESCRIPTION
**Pull request by BeyonderXX**
## Description ##
I did hybridize( ) and pass "valid_length" to the unroll( ) function of BidirectionalCell, then returned AssertionError in line 79. Because symbol.split( ) return a symbol but not a symbol list. Result in the length of inputs is not equal to parameter "length"  when call unroll( )  to compute r_outputs and r_states. Symbol support for converting to list through list( ) API, so that use list( ) function insted of _as_list( ).

### Essentials
Please feel free to remove inapplicable items for your PR.

* [x]  Changes are complete (i.e. I finished coding on this PR)
* [x]  All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
-  Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
 * [x]  To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
python/mxnet/gluon/rnn/rnn_cell.py
- [line 1055] [line 1069]add _reverse_sequences( ) to calculate "reversed_inputs" and "reversed_r_outputs" 
- [line 1055] use list( ) instead of _as_list( )

tests/python/unittest/test_gluon_rnn.py
- [line 603] add test_bidirectional_unroll_valid_length( ) to test the unroll( ) of BidirectionalCell when pass 
valid_length to it.

### CODE REVIEW

**Comment by szha** [comment]
@larroy _as_list would return a wrong list as it would wrap the single grouped symbol inside, instead of `length` number of tensors for each time-step, which is what the output should be.

**Comment by szha** [comment]
@BeyonderXX thanks for the fix. Could you add a test with the proper assertions?

**Comment by BeyonderXX** [comment]
> @BeyonderXX thanks for the fix. Could you add a test with the proper assertions?

I‘m glad to. Do you mean add assertion to this code or provide a test case with assertion? If the latter case, is there any sample test?

File: tests/python/unittest/test_gluon_rnn.py [hunk 1]
```diff
@@ -243,6 +243,34 @@ def test_bidirectional():
...
+            self.time_step = time_step
+            with self.name_scope():
+                self.bi_lstm = gluon.rnn.BidirectionalCell(
+                    gluon.rnn.LSTMCell(rnn_size, prefix='rnn_l0_'),
+                    gluon.rnn.LSTMCell(rnn_size, prefix='rnn_r0_'),
+                    output_prefix='lstm_bi_')
+
+        def hybrid_forward(self, F, inputs, valid_len):
+            outputs, states = self.bi_lstm.unroll(self.time_step, inputs, valid_length=valid_len,
+                                                  layout='TNC', merge_outputs='True')
```
Comment (szha) [inline tests/python/unittest/test_gluon_rnn.py]: I think merge_outputs should be boolean or None

Comment (BeyonderXX) [reply tests/python/unittest/test_gluon_rnn.py]: Thank you so much!

**Comment by BeyonderXX** [comment]
@szha I just added one test case named "test_bidirectional_unroll_valid_length" to test_gluon_rnn. While jenkins detect another test called "test_layer_bidirectional " which returned error in various plateform while using cpu. Is this OK or raise by my code in BidirectionalCell ? But jenkins checked pass at first version. 
This is error info,
"test_gluon_rnn.test_bidirectional_unroll_valid_length ... ok (0.0268s)

test_gluon_rnn.test_layer_bidirectional ... *** Error in `python3.6': munmap_chunk(): invalid pointer: 0x00007ff25c1f3140 ***
"

**Comment by BeyonderXX** [comment]
@szha I reproduced the issue in windows. Might unroll( ) would raise null pointer when pass valid_length after hybridize( ).

**Comment by BeyonderXX** [comment]
@szha  I figured out the issue, because I passed a wrong parameter to unroll( ) which led to out of bounds. I have fixed the problem and rewrited my code in BidirectionalCell.  So sorry to disturb you abruptly. Thanks!

**Comment by szha** [comment]
@BeyonderXX thanks for the fix!

**Pull request merged**